include/envoy/upstream/cluster_manager.h

```diff
@@ -147,6 +170,17 @@ class ClusterManager {
    * @return std::string the local cluster name, or "" if no local cluster was configured.
    */
   virtual const std::string& localClusterName() const PURE;
+
+  /**
+   * These two methods allow to register and unregister callbacks for cluster lifecycle events
+   * in the ClusterManager.
+   * The callbacks will be registered in a thread local slot and the callbacks will be executed
+   * on the thread that registered them.
+   *
+   * @param callbacks are the ClusterUpdateCallbacks to add or remove to the cluster manager.
+   */
+  virtual void addClusterUpdateCallbacks(ClusterUpdateCallbacks& callbacks) PURE;
+  virtual void removeClusterUpdateCallbacks(ClusterUpdateCallbacks& callbacks) PURE;
```

Comment:
One quick question on API. Given that we expect to potentially add/remove these frequently, and will have to deal with filter lifecycle events, should we have an RAII pattern here where `add` returns a handle that implicitly does a remove when it is destroyed?

Reply:
I am not versed enough in cpp so I'll let others comment on this one.